app/controllers/api_controller.rb

```diff
@@ -3,6 +3,52 @@ class ApiController < ApplicationController
 
   private
 
+  ##
+  # Set default request format to xml unless a client requests a specific format,
+  # which can be done via (a) URL suffix and/or (b) HTTP Accept header, where
+  # the URL suffix always takes precedence over the Accept header.
+  def set_default_request_format
+    unless params[:format]
+      accept_header = request.headers["HTTP_ACCEPT"]
+      if accept_header.nil?
+        # e.g. unit tests don't set an Accept: header by default, force XML in this case
+        request.format = "xml"
+        return
+      end
+
+      req_mimetypes = []
+
+      # Some clients (JOSM) send Accept headers which cannot be parsed by Rails, example: *; q=.2
+      # To be fair, JOSM's Accept header doesn't adhere to RFC 7231, section 5.3.1, et al. either
+      # As a workaround for backwards compatibility, we're assuming XML format
+      begin
+        req_mimetypes = Mime::Type.parse(accept_header)
+      rescue Mime::Type::InvalidMimeType
+        request.format = "xml"
+        return
+      end
+
+      # req_mimetypes contains all Accept header MIME types with descending priority
+      req_mimetypes.each do |mime|
+        if mime.symbol == :xml
+          request.format = "xml"
+          break
+        end
+
+        if mime.symbol == :json
+          request.format = "json"
+          break
+        end
+
+        # Any format, not explicitly requesting XML or JSON -> assume XML as default
+        if mime == "*/*"
+          request.format = "xml"
+          break
+        end
+      end
```

Comment:
TODO: Add a comment here, that we deliberately don't set request.format, in case not XML, JSON or */* was requested, so that an HTTP error is returned back to the client.

Reply:
But this is a breaking change to the API, right? Now, if you request a resource with an Accept header of "text/plain" then you get the XML response. This PR would change that behaviour. That's why I discussed having the default being XML.

Of course, I would propose to fix this in API 0.7 and behave more sensibly wrt Accept headers in the new API version, but that will need to wait until there's progress with that topic.

Reply:
That's how the future behaviour looks like vs. what we have on osm.org/master.apis:

`curl -v 'https://master.apis.dev.openstreetmap.org/api/0.6/map?bbox=-61.083984375,45.5371366803986,-61.07849121093749,45.540984218050774' -H "Accept: text/plain`

Returns: HTTP 406 (also for www.openstreetmap.org).

`curl -v 'http://localhost:3000/api/0.6/map?bbox=-61.083984375,45.5371366803986,-61.07849121093749,45.540984218050774' -H "Accept: text/plain"`

500 Internal Server Error - ActionController::UnknownFormat: ActionController::UnknownFormat

Reply:
Ah, is because cgimap is doing this? I have different results for e.g. my own localhost, or other rails port deployments like OpenHistoricalMap

I think this might be where cgimap is doing so, based on the error messages:
https://github.com/zerebubuth/openstreetmap-cgimap/blob/ee2d8078db851505888ddad275419066a4015cef/src/choose_formatter.cpp#L236

So it is a change in behaviour for the rails port, but given that cgimap already does it's own thing here, it's less of an issue.

Reply:
In both cases the response was generated by cgimap, so yes, Rails port and cgimap behaved differently for many years here.

In general, Accept header support in Rails is quite a delicate topic. There's a number of issues out there, like adding `*/*` would break multi format output, or as in your example, requesting text/plain would happily return XML.

It was the main reason for another round of Accept header parsing in _set_default_request_format_, rather than relying on the framework to do the right thing.